Implemented initial Django Channels setup #47

HlibSokol opened the pull request:
- installed _channels_ and _daphne_
- added _channels_ and _daphne_ to requirements.txt and to INSTALLED_APPS in settings.py
- added ASGI_APPLICATION in settings.py
- configured asgi.py with ProtocolTypeRouter for "http" and "websocket"
- created routing.py to handle websocket paths
- successfully tested connection using Postman
![image](https://github.com/user-attachments/assets/0b5ef0cd-cdd0-449b-b450-468ea9f61339)

VovaD199 commented on forum/forum/settings.py:
SIMPLE_JWT specifies SIGNINS_KEY = SECRET_KEY — looks like an error, it should be SIGNING_KEY.

HlibSokol replied:
It is already fixed in last PR (JWT authentication)
![image](https://github.com/user-attachments/assets/7c7380fe-34b8-463d-a64f-e51ef02e6ce8)

VovaD199 commented on forum/forum/routing.py:
```diff
@@ -0,0 +1,24 @@
+    path("ws/test/", Consumer.as_asgi()),
+]
+
+application = URLRouter(websocket_urlpatterns)    
```
Invalid application object
The routing.py file **should not export URLRouter directly as application**. It should be **part of** the ProtocolTypeRouter in asgi.py, as you already did there:
application = ProtocolTypeRouter({
    "http": get_asgi_application(),
    "websocket": URLRouter(websocket_urlpatterns)
})

VovaD199 commented on forum/forum/routing.py:
You are not using AuthMiddlewareStack, so WebSocket sessions do not have access to the user (self.scope["user"] will be anonymous).

from channels.auth import AuthMiddlewareStack

websocket_urlpatterns = [
    path("ws/test/", AuthMiddlewareStack(Consumer.as_asgi())),
]


This is required if you plan to authorize WebSocket users.

HlibSokol replied:
I added AuthMiddlewareStack to asgi.py so now all users/consumers will be authorized on ASGI level - better than adding AuthMiddlewareStack to every route in routing.py

HlibSokol commented:
- updated doc-string in asgi.py with comprehensive description of structure
- added AuthMiddlewareStack to asgi.py so now WebSocket users can be authorized and are not anonymous 
- added try/except to imports for secure check
- removed invalid 'application' object from routing.py
- added second parameter to 'async def disconnect(self, **_code_**)' in Consumer so now it disconnects without TypeError (because it expects 2 arguments instead of 1)

mehalyna commented on forum/forum/asgi.py:
```diff
@@ -5,12 +5,49 @@
 
 For more information on this file, see
 https://docs.djangoproject.com/en/4.2/howto/deployment/asgi/
+
```
Add a brief comment so future developers understand this is not just Django's default ASGI but supports both HTTP and WebSockets:
```
# ASGI application supporting both HTTP and WebSocket protocols via Channels
```

mehalyna commented on forum/forum/routing.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class Consumer(AsyncWebsocketConsumer):
+    async def connect(self):
+        await self.accept()
+        await self.send(text_data="Websocket connected.")
```
Use `self.scope["user"]` if you're using `AuthMiddlewareStack`
This helps you validate authenticated users:
```
    async def connect(self):
        user = self.scope["user"]
        if user.is_authenticated:
            await self.accept()
            await self.send(text_data="WebSocket connected.")
        else:
            await self.close()
```
Makes the connection conditional, avoiding unauthorized access in future.

mehalyna commented on forum/forum/routing.py:
```diff
@@ -0,0 +1,22 @@
+    async def receive(self, text_data):
+        await self.send(text_data=f"You sent: {text_data}")
+
+    async def disconnect(self, code):
+        pass
```
Currently, `disconnect` is a no-op. At minimum, log the disconnect code:
```
    async def disconnect(self, code):
        print(f"WebSocket disconnected with code: {code}")
```
This helps debug client disconnects or network failures.

HlibSokol commented:
now only authenticated users can be connected via WebSocket
![image](https://github.com/user-attachments/assets/c56cff9f-eefa-4a62-a3aa-bcb648849695)
